app/src/main/cpp/skyline/nce.cpp

```diff
@@ -1,231 +1,319 @@
 #include <sched.h>
-#include <linux/uio.h>
-#include <linux/elf.h>
+#include <unistd.h>
 #include "os.h"
 #include "jvm.h"
+#include "nce/guest.h"
+#include "nce/instr.h"
+#include "kernel/svc.h"
+#include "nce.h"
 
 extern bool Halt;
-extern std::mutex jniMtx;
+extern jobject Surface;
+extern skyline::GroupMutex jniMtx;
 
 namespace skyline {
-    void NCE::ReadRegisters(user_pt_regs &registers, pid_t pid) const {
-        iovec iov = {&registers, sizeof(registers)};
-        long status = ptrace(PTRACE_GETREGSET, pid ? pid : currPid, NT_PRSTATUS, &iov);
-        if (status == -1)
-            throw exception("Cannot read registers, PID: {}, Error: {}", pid, strerror(errno));
+    void NCE::KernelThread(pid_t thread) {
+        try {
+            state.thread = state.process->threads.at(thread);
+            state.ctx = reinterpret_cast<ThreadContext *>(state.thread->ctxMemory->kernel.address);
+            while (true) {
+                asm("yield");
```

Comment:
Do we really want to yield here? Probably an idea to benchmark this first.

Reply:
We do, this essentially spins doing nothing. That tends to just waste a lot of CPU resources, a yield will redirect energy towards other cores which can utilize it better hopefully.